File: src/clients/mix-server-client/android-plugin/nativeaudio/NativeAudioPlugin.kt
package com.mixserver.app.plugins.nativeaudio

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "NativeAudio")
class NativeAudioPlugin : Plugin() {

    private lateinit var audioPlayer: NativeAudioPlayer

    override fun load() {
        audioPlayer = NativeAudioPlayer(context, this)
    }

    @PluginMethod
    fun setSource(call: PluginCall) {
        val url = call.getString("url")
        if (url == null) {
            call.reject("Missing 'url' parameter")
            return
        }
        audioPlayer.setSource(url)
        call.resolve()
    }

    @PluginMethod
    fun play(call: PluginCall) {
        val fromTime = call.getDouble("fromTime")
        audioPlayer.play(fromTime)
        call.resolve()
    }

    @PluginMethod
    fun pause(call: PluginCall) {
        audioPlayer.pause()
        call.resolve()
    }

    @PluginMethod
    fun seek(call: PluginCall) {
        val time = call.getDouble("time")
        if (time == null) {
            call.reject("Missing 'time' parameter")
            return
        }
        audioPlayer.seek(time)
        call.resolve()
    }

    @PluginMethod
    fun stop(call: PluginCall) {
        audioPlayer.stop()
        call.resolve()
    }

    @PluginMethod
    fun getCurrentTime(call: PluginCall) {
        val result = JSObject()
        result.put("currentTime", audioPlayer.currentTime)
        call.resolve(result)
    }

    @PluginMethod
    fun getDuration(call: PluginCall) {
        val result = JSObject()
        result.put("duration", audioPlayer.duration)
        call.resolve(result)
    }

    @PluginMethod
    fun setMetadata(call: PluginCall) {
        val title = call.getString("title") ?: ""
        val artist = call.getString("artist")
        val album = call.getString("album")
        val duration = call.getDouble("duration")
        audioPlayer.setMetadata(title, artist, album, duration)
        call.resolve()
    }

    @PluginMethod
    fun setNowPlayingPosition(call: PluginCall) {
        val position = call.getDouble("position")
        val duration = call.getDouble("duration")
        val playbackRate = call.getDouble("playbackRate")
        if (position == null || duration == null || playbackRate == null) {
            call.reject("Missing required parameters")
            return
        }
        audioPlayer.setNowPlayingPosition(position, duration, playbackRate)
        call.resolve()
    }

    @PluginMethod
    fun setSkipControls(call: PluginCall) {
        val hasNext = call.getBoolean("hasNext") ?: false
        val hasPrevious = call.getBoolean("hasPrevious") ?: false
        audioPlayer.setSkipControls(hasNext, hasPrevious)
        call.resolve()
    }

    // Event forwarding from NativeAudioPlayer

    fun notifyPlayRequest() {
        notifyListeners("playRequest", JSObject())
    }

    fun notifyPauseRequest() {
        notifyListeners("pauseRequest", JSObject())
    }

    fun notifySeekRequest(time: Double) {
        val data = JSObject()
        data.put("time", time)
        notifyListeners("seekRequest", data)
    }

    fun notifyNextTrackRequest() {
        notifyListeners("nextTrackRequest", JSObject())
    }

    fun notifyPreviousTrackRequest() {
        notifyListeners("previousTrackRequest", JSObject())
    }

    fun notifyTimeUpdate(currentTime: Double) {
        val data = JSObject()
        data.put("currentTime", currentTime)
        notifyListeners("timeUpdate", data)
    }

    fun notifyDurationChange(duration: Double) {
        val data = JSObject()
        data.put("duration", duration)
        notifyListeners("durationChange", data)
    }

    fun notifyEnded() {
        notifyListeners("ended", JSObject())
    }

    fun notifyError(message: String) {
        val data = JSObject()
        data.put("message", message)
        notifyListeners("error", data)
    }

    override fun handleOnDestroy() {
        audioPlayer.release()
        super.handleOnDestroy()
    }
}

File: src/clients/mix-server-client/ios-plugin/NativeAudio/NativeAudioPlugin.swift
import Foundation
import Capacitor

/// Capacitor plugin bridge for native audio playback.
/// Register in the app's AppDelegate or via the Capacitor plugin loader.
@objc(NativeAudioPlugin)
public class NativeAudioPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NativeAudioPlugin"
    public let jsName = "NativeAudio"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setSource", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "play", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pause", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "seek", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stop", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCurrentTime", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getDuration", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setMetadata", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setNowPlayingPosition", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setSkipControls", returnType: CAPPluginReturnPromise),
    ]

    private lazy var player: NativeAudioPlayer = {
        return NativeAudioPlayer(plugin: self)
    }()

    @objc func setSource(_ call: CAPPluginCall) {
        guard let url = call.getString("url") else {
            call.reject("Missing 'url' parameter")
            return
        }
        player.setSource(urlString: url)
        call.resolve()
    }

    @objc func play(_ call: CAPPluginCall) {
        let fromTime = call.getDouble("fromTime")
        player.play(fromTime: fromTime)
        call.resolve()
    }

    @objc func pause(_ call: CAPPluginCall) {
        player.pause()
        call.resolve()
    }

    @objc func seek(_ call: CAPPluginCall) {
        guard let time = call.getDouble("time") else {
            call.reject("Missing 'time' parameter")
            return
        }
        player.seek(to: time)
        call.resolve()
    }

    @objc func stop(_ call: CAPPluginCall) {
        player.stop()
        call.resolve()
    }

    @objc func getCurrentTime(_ call: CAPPluginCall) {
        call.resolve(["currentTime": player.currentTime])
    }

    @objc func getDuration(_ call: CAPPluginCall) {
        call.resolve(["duration": player.duration])
    }

    @objc func setMetadata(_ call: CAPPluginCall) {
        let title = call.getString("title") ?? ""
        let artist = call.getString("artist")
        let album = call.getString("album")
        let duration = call.getDouble("duration")
        player.setMetadata(title: title, artist: artist, album: album, duration: duration)
        call.resolve()
    }

    @objc func setNowPlayingPosition(_ call: CAPPluginCall) {
        guard let position = call.getDouble("position"),
              let duration = call.getDouble("duration"),
              let playbackRate = call.getDouble("playbackRate") else {
            call.reject("Missing required parameters")
            return
        }
        player.setNowPlayingPosition(position: position, duration: duration, playbackRate: playbackRate)
        call.resolve()
    }

    @objc func setSkipControls(_ call: CAPPluginCall) {
        let hasNext = call.getBool("hasNext") ?? false
        let hasPrevious = call.getBool("hasPrevious") ?? false
        player.setSkipControls(hasNext: hasNext, hasPrevious: hasPrevious)
        call.resolve()
    }

    // MARK: - Event forwarding from NativeAudioPlayer

    func notifyPlayRequest() {
        notifyListeners("playRequest", data: [:])
    }

    func notifyPauseRequest() {
        notifyListeners("pauseRequest", data: [:])
    }

    func notifySeekRequest(time: Double) {
        notifyListeners("seekRequest", data: ["time": time])
    }

    func notifyNextTrackRequest() {
        notifyListeners("nextTrackRequest", data: [:])
    }

    func notifyPreviousTrackRequest() {
        notifyListeners("previousTrackRequest", data: [:])
    }

    func notifyTimeUpdate(currentTime: Double) {
        notifyListeners("timeUpdate", data: ["currentTime": currentTime])
    }

    func notifyDurationChange(duration: Double) {
        notifyListeners("durationChange", data: ["duration": duration])
    }

    func notifyEnded() {
        notifyListeners("ended", data: [:])
    }

    func notifyError(message: String) {
        notifyListeners("error", data: ["message": message])
    }
}
